File: 46-permutations/46-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    
    function _permute( candidates, perm) {
        
        if (candidates.length === 0) {
            permutations.push([...perm])
            return
        }
        
        for (let cand of candidates) {
            perm.push(cand);
            _permute(
                candidates.filter(elem => elem != cand),
                perm
            )
            perm.pop()
        }
    }
    
    var permutations = []
    _permute(nums, []);
    return permutations;    
};

File: 46-permutations/46-permutations.java
class Solution {
    
    List<List<Integer>> permutations = new ArrayList<>();
    
    public List<List<Integer>> permute(int[] nums) {
        ////////////////////////////////////////////
        List<Integer> numList = Arrays.stream(nums).boxed().collect(Collectors.toList());
        _permute(
            numList,
            new ArrayList<Integer>()
        );
        return permutations;
    }
    
    public void _permute(List<Integer> candidates, List<Integer> perm) {
        
        if (candidates.size() == 0) {
            //////////////////////////////////////
            permutations.add( new ArrayList<>(perm) );
            return;
        }
        
        for (Integer cand : candidates ) {
            perm.add(cand);
            _permute(
                _filterList(candidates, cand),
                perm
            );
            perm.remove(perm.size() - 1);
        }
    }
    
    public List<Integer> _filterList(List<Integer> list, Integer elemToRemove) {
        //////////////////////////////////////
        List<Integer> copy = new ArrayList<>(list);
        copy.removeIf(elem -> elem == elemToRemove);
        return copy;
    }
}